Added support for Crate

chaudum opened the pull request:
Implements #2200 

Hi there!

This PR adds basic support for [Crate](https://crate.io) as a database backend for Metabase. We know that certain tests fail, etc. but we opened the PR to start the discussion what is still to do and how we can overcome certain problems we're facing during implementation of the driver.

Let's get this thing done! 🚀
###### TODO
- [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)

Commit:
register Crate in Metabase

Commit:
added crate dependencies

Commit:
added verification for Crate connection setup

Commit:
added base field types for demo table

Commit:
fix: changed incorrect datatype mapping

Commit:
fix: added char_length function handle

Commit:
applied custom filter

Commit:
add test environment

aslanbakirov committed:
Added Crate type mappings

aslanbakirov committed:
Fixup! Added Crate type mappings

Commit:
implement date time handling

Commit:
code formatting

Commit:
added support for objects and arrays

Commit:
fixed exceptions if database gets added

* caused by incorrect cast if avg field length gets calculated

Commit:
extend generic sql to avoid disable auto-commit

 - restructered driver package
 - use Crate specific implementation on raw sql queries

Commit:
use JDBC library to disable autocommit in test

Commit:
fix primary key sql type

Commit:
load test-data parses to valid Crate timestamp

Commit:
changed crate input paramters so it’s possible
to connect to multiple hosts

Commit:
fix: set default schema to `doc`

Commit:
fix: added supported driver features

fixes some tests of query-processor

chaudum committed:
rewrite between filter

Commit:
fix test that checks date-interval implementation

chaudum committed:
map Crate data types correctly

chaudum committed:
fixed broken imports after rebase with upstream master

chaudum commented on src/metabase/driver/crate.clj:
```diff
@@ -0,0 +1,96 @@
+    :ip               :TextField
+    :timestamp        :DateTimeField
+    :geo_shape        :DictionaryField
+    :geo_point        :ArrayField
```
the `GEO_POINT` data type is an array of `[ <longitude>, <latitude> ]`
maybe there is a better way to display it instead of an array, e.g. as separate "virtual columns" for lat and lon which then may be used for the pin map visualisation

camsaul replied:
@chaudum Metabase doesn't currently directly support array column data types so for the time being leaving it as-is is fine.

chaudum commented on src/metabase/driver/crate.clj:
```diff
@@ -0,0 +1,96 @@
+  (:import (clojure.lang Named)))
+
+(defn- column->base-type
+  "Map of Crate column types -> Field base types"
```
for reference: Crate data types - https://crate.io/docs/reference/sql/data_types.html

camsaul replied:
I would also put this link in a comment or docstring in the code itself

chaudum commented on src/metabase/driver/crate/native.clj:
```diff
@@ -0,0 +1,45 @@
+            [clojure.tools.logging :as log]
+            [metabase.util :as u]))
+
+(defn- value->base-type
```
this function definition is duplicate in our driver, because the original is private
some times we were not sure, whether to duplicate code, or change the definition to be public

camsaul replied:
it's better to change the definition to be public

chaudum commented on src/metabase/driver/crate/query_processor.clj:
```diff
@@ -0,0 +1,18 @@
+  "Apply custom generic SQL filter. This is the place to perform query rewrites."
+  [_ korma-form {clause :filter}]
+  (k/where korma-form (case (:filter-type clause)
+            :between (qp/filter-clause->predicate (rewrite-between clause))
```
since Crate does not support the `BETWEEN` predicate, we had to rewrite it to `<min> <= <value> <= <max>`
not sure if this is the correct place, it was definitely the easiest, though ;)

chaudum commented on src/metabase/driver/generic_sql.clj:
```diff
@@ -85,6 +86,9 @@
     "Return a korma form appropriate for converting a Unix timestamp integer field or value to an proper SQL `Timestamp`.
      SECONDS-OR-MILLISECONDS refers to the resolution of the int in question and with be either `:seconds` or `:milliseconds`."))
 
+(extend-protocol jdbc/IResultSetReadColumn
```
I think, extending the protocol has effect on ALL jdbc drivers. Could not find another way, though, to correctly deal with the `Object[]` object from the jdbc result set.
This will display Crate's [`ARRAY` type](https://crate.io/docs/reference/sql/data_types.html#array) correctly, if it is a "root column".
`ARRAY` inside an `OBJECT` column is still displayed as `[Ljava.langObject@...` - I think the `Object` type from the database needs to be converted recursively.

agilliland replied:
I'll play around with this a little.  You are right that this will impact all drivers since it applies to the JDBC interface as a whole.

chaudum commented on src/metabase/driver/generic_sql/query_processor.clj:
```diff
@@ -167,7 +167,7 @@
              :!=          ['not= (formatted value)])}))
 
-(defn- filter-clause->predicate [{:keys [compound-type subclause subclauses], :as clause}]
+(defn filter-clause->predicate [{:keys [compound-type subclause subclauses], :as clause}]
```
we wanted to reuse the function ...

Commit:
code cleanup

Commit:
adapt circle-ci configuration file to use Crate

Commit:
fix several date-interval issues

agilliland commented on src/metabase/driver/crate/native.clj:
```diff
@@ -0,0 +1,31 @@
+            [metabase.util :as u]
+            [metabase.driver.generic-sql.native :as n]))
+
+(defn process-and-run
```
what is the reason behind duplicating this function instead of using the one in `generic-sql/native` ?

mikethebeer replied:
Crate does not support transactions. Therefore we need to use the the `jdbc/with-db-connection` instead of `jdbc/with-db-transaction` which sets autocommit to `false` - which is not good for us 😢 since our JDBC driver is returning a `SQLFeatureNotSupportedException` https://github.com/crate/crate-jdbc/blob/1.10/src/main/java/io/crate/client/jdbc/CrateConnection.java#L90.

agilliland replied:
got it.  how about the timezone support though, do you guys provide a way to set the connection timezone?

mikethebeer replied:
@agilliland no not explicitly for the connection like postgresql does. However, in the `date_trunc` scalar function we are supporting timezones (see https://crate.io/docs/reference/sql/scalar.html#date-trunc-interval-timezone-timestamp-returns-timestamp). If its not explicitly specified it is always based on UTC - like it is now.
I will discuss that internally if we provide it there or not.

mikethebeer replied:
@agilliland implemented timezone-support (see https://github.com/metabase/metabase/pull/2428/commits/d9d95ad273c598d6113ccb08e0c2454dfadf7540)

Commit:
implement inside filter

mikethebeer commented:
Hi @camsaul ,
the driver is now ready on our side. Integration-Tests are passing, however we don't fix 2 failures which we receive from extracting parts of a timestamp. Its a bit odd that Metabase is testing those database specific implementations explicitly since those are already tested on our side. Although it is not clear for us where such queries containing `days_of_week` or `week` as date-interval can be set-up using the UI.

So we keep those tests as they are - as an assumption that each database has its own implementation of extracting parts out of a timestamp.

agilliland commented:
Hi @mikethebeer , the reason we test all of these details is mainly to ensure that across all of the databases we can see consistent output, so i think we should dig in a little bit more and understand why those tests are failing.

If you want to test them out in the UI you can click `New Question` at the top of the app, then pick a table which has a timestamp column in it, choose `Count of Rows` and then under `Add a Grouping` you'll scroll down to your timestamp column and click on the right side by the arrow to expand a second menu which allows you to group the data by various time granularities and extracts.

agilliland commented on circle.yml:
```diff
@@ -32,7 +36,8 @@ test:
     # 6) runs lein uberjar. (We don't run bin/build because we're not really concerned about `npm install` (etc) in this test, which runs elsewhere)
-    - case $CIRCLE_NODE_INDEX in 0) ENGINES=h2,mongo,mysql,bigquery lein test ;; 1) ENGINES=h2,sqlserver MB_DB_TYPE=postgres MB_DB_DBNAME=circle_test MB_DB_PORT=5432 MB_DB_USER=ubuntu MB_DB_HOST=localhost lein test ;; 2) ENGINES=h2,postgres,sqlite MB_DB_TYPE=mysql MB_DB_DBNAME=circle_test MB_DB_PORT=3306 MB_DB_USER=ubuntu MB_DB_HOST=localhost lein test ;; 3) ENGINES=h2,redshift,druid lein test ;; ...
+    # 7) runs unit tests w/ H2 local DB. Runs agains H2, Crate
+    - case $CIRCLE_NODE_INDEX in 0) ENGINES=h2,mongo,mysql,bigquery lein test ;; 1) ENGINES=h2,sqlserver MB_DB_TYPE=postgres MB_DB_DBNAME=circle_test MB_DB_PORT=5432 MB_DB_USER=ubuntu MB_DB_HOST=localhost lein test ;; 2) ENGINES=h2,postgres,sqlite MB_DB_TYPE=mysql MB_DB_DBNAME=circle_test MB_DB_PORT=3306 MB_DB_USER=ubuntu MB_DB_HOST=localhost lein test ;; 3) ENGINES=h2,redshift,druid lein test ;; ...
```
we don't actually want to run a 7th container here, instead we should add create to 1) which is only testing h2 and sqlserver right now.

mikethebeer replied:
@agilliland okay understand. I created a 7th container in order to prevent from getting database credential and authorisation errors if no settings are configured for SQLServer or others.

camsaul commented on src/metabase/driver/crate/query_processor.clj:
```diff
@@ -0,0 +1,18 @@
+  "Rewrite [:between <field> <min> <max>] -> [:and [:>= <field> <min>] [:<= <field> <max>]]"
+  [clause]
+  (CompoundFilter. :and [(ComparisonFilter. :>= (:field clause) (:min-val clause))
+                         (ComparisonFilter. :<= (:field clause) (:max-val clause))]))
```
instead of using the constructor directly `CompoundFilter.` it's preferable to use the `map->CompoundFilter` or `strict-map->CompoundFilter` functions because they're more readable and future-proof

camsaul commented on src/metabase/driver/crate/util.clj:
```diff
@@ -0,0 +1,77 @@
+    :hour (k/raw (sql-interval HOUR amount))
+    :minute (k/raw (sql-interval MINUTE amount))
+    :second (k/raw (sql-interval SECOND amount))
+    ))
```
In Clojure closing parens don't normally get a new line

camsaul commented on src/metabase/driver/crate/util.clj:
```diff
@@ -0,0 +1,77 @@
+(def ^:private DAY    (constantly 86400000))
+(def ^:private HOUR   (constantly 3600000))
+(def ^:private MINUTE (constantly 60000))
+(def ^:private SECOND (constantly 1000))
```
in Clojure constants aren't uppercased. And since these values are constants there's not much sense in making them functions. I would also suggest composing them for clarity, e.g.:

```clojure
(def ^:private ^:const second 1000)
(def ^:private ^:const minute (* 60 second))
(def ^:private ^:const hour   (* 60 minute))


(defn- sql-interval [unit amount]
  (format "CURRENT_TIMESTAMP + %d" (* unit amount)))
```

camsaul commented on src/metabase/driver/generic_sql.clj:
```diff
@@ -85,6 +86,9 @@
 
+(extend-protocol jdbc/IResultSetReadColumn
+  (class (object-array []))
+  (result-set-read-column [x _2 _3] (PersistentVector/adopt x)))
```
In Clojure arguments that are ignored are just called `_`. So this should be something like

```clojure
(result-set-read-column [x _ _] (PersistentVector/adopt x))
```

camsaul commented on test/metabase/query_processor_test.clj:
```diff
@@ -780,15 +780,15 @@
   {:columns [(format-name "price")
              "stddev"]
-   :rows    [[3 (if (= *engine* :mysql) 25 26)]
+   :rows    [[3 (if (or (= *engine* :mysql) (= *engine* :crate)) 25 26)]
```
You should probably just do
```clojure
(contains? #{:mysql :crate} *engine*)
```
instead

camsaul commented on test/metabase/test/data/crate.clj:
```diff
@@ -0,0 +1,75 @@
+            [metabase.driver.generic-sql :as sql]
+            [clojure.java.jdbc :as jdbc]
+            [metabase.util :as u]
+            [clojure.string :as s])
```
Do you mind ordering the `:requires`? In Metabase we keep the the dependent namespaces ordered like:

*  Clojure namespaces
*  3rd-party namespaces
*  metabase namespaces

See other MB namespaces to get a feel for how we do it.

camsaul commented on test/metabase/test/data/crate.clj:
```diff
@@ -0,0 +1,75 @@
+
+(defn- timestamp->CrateDateTime
+  [value]
+  (if (= (instance? java.sql.Timestamp value) true)
```
no need to do `(= ... true)` here.

Write it like this instead:

```clojure
(if (instance? java.sql.Timestamp value)
  ...)
```

camsaul commented on test/metabase/test/data/crate.clj:
```diff
@@ -0,0 +1,75 @@
+
+(defn- database->connection-details [_ _ {:keys [_ _]}]
+  (merge {:host         "localhost"
+          :port         4300}))
```
If these will indeed never change this should be written like

```clojure
(def ^:private database->connection-details
  (constantly {:host "localhost"
               :port 4300}))
```

camsaul commented on test/metabase/test/data/crate.clj:
```diff
@@ -0,0 +1,75 @@
+    (let [insert!   ((apply comp wrap-insert-fns) (fn [row-or-rows]
+                                                    (apply jdbc/insert!
+                                                           (generic/database->spec driver :db dbdef)
+                                                           (keyword (get tabledef :table-name))
```
No need to use `get` here. `(:table-name tabledef)` means the same thing and is more concise.

camsaul commented on test/metabase/test/data/crate.clj:
```diff
@@ -0,0 +1,75 @@
+  (if (= (instance? java.sql.Timestamp value) true)
+    (.getTime (u/->Timestamp value))
+    (if (= (and (= (instance? clojure.lang.PersistentArrayMap value) true) (contains? value :korma.sql.utils/generated)) true)
+      (+ (read-string (s/replace (:korma.sql.utils/generated value) #"CURRENT_TIMESTAMP \+" "")) (.getTime (u/new-sql-timestamp)))
```
You should be using `(korma.core/raw value)` instead of constructing that function's results yourself like `{:korma.sql.utils value}`

mikethebeer replied:
@camsaul hm. I know this line looks really odd. However, I wasn't able to use the `korma.core/raw` function to generate the insert-value. I think it is related to plain `jdbc/insert!` method that I use to generate a non-transactional insert (Crate does not support transactions). I'm using it now like this:

```clojure
(defn- timestamp->CrateDateTime
  [value]
  (if (instance? java.sql.Timestamp value)
    (.getTime (u/->Timestamp value))
    (if (and (instance? clojure.lang.PersistentArrayMap value) (contains? value :korma.sql.utils/generated))
      (k/raw value)
      value)))
```

And get this stacktrace:
```
act-msg: exception in actual: (metabase.test.data.datasets/do-with-engine :crate actual--873848585)
    threw: class java.lang.ClassCastException - clojure.lang.Keyword cannot be cast to java.lang.String
           io.crate.shade.org.elasticsearch.common.io.stream.StreamOutput$writeGenericValue (StreamOutput.java:381)
           io.crate.action.sql.SQLRequest$writeTo (SQLRequest.java:96)
           io.crate.shade.org.elasticsearch.transport.netty.NettyTransport$sendRequest (NettyTransport.java:701)
           io.crate.shade.org.elasticsearch.transport.TransportService$sendRequest (TransportService.java:276)
           io.crate.shade.org.elasticsearch.action.TransportActionNodeProxy$execute (TransportActionNodeProxy.java:57)
           io.crate.client.InternalCrateClient$1$doWithNode (InternalCrateClient.java:92)
           io.crate.shade.org.elasticsearch.client.transport.TransportClientNodesService$execute (TransportClientNodesService.java:206)
           io.crate.client.InternalCrateClient$execute (InternalCrateClient.java:87)
           io.crate.client.InternalCrateClient$execute (InternalCrateClient.java:70)
           io.crate.client.InternalCrateClient$sql (InternalCrateClient.java:59)
           io.crate.client.CrateClient$sql (CrateClient.java:151)
           io.crate.client.jdbc.CratePreparedStatement$executeSingle (CratePreparedStatement.java:173)
           io.crate.client.jdbc.CratePreparedStatement$execute (CratePreparedStatement.java:163)
           io.crate.client.jdbc.CratePreparedStatement$executeUpdate (CratePreparedStatement.java:124)
           on (jdbc.clj:727)
           clojure.java.jdbc$db_do_prepared_return_keys$invokeStatic (jdbc.clj:742)
           on (jdbc.clj:714)
           on (jdbc.clj:937)
           clojure.java.jdbc$multi_insert_helper$invokeStatic (jdbc.clj:936)
           on (jdbc.clj:931)
           clojure.java.jdbc$insert_helper$invokeStatic (jdbc.clj:948)
           on (jdbc.clj:940)
           clojure.java.jdbc$insert_BANG_$invokeStatic (jdbc.clj:1039)
           clojure.java.jdbc$insert_BANG_$doInvoke (jdbc.clj:1024)
           on (crate.clj:48)
           on (generic_sql.clj:183)
           on (crate.clj:55)
           on (generic_sql.clj:22)
           metabase.test.data.generic_sql$create_db_BANG_$invokeStatic (generic_sql.clj:324)
           on (generic_sql.clj:300)
           on (interface.clj:64)
           metabase.test.data$get_or_create_database_BANG_$invokeStatic (data.clj:178)
           on (data.clj:167)
           metabase.test.data$do_with_temp_db$invokeStatic (data.clj:248)
           on (data.clj:243)
           metabase.query_processor_test$count_of_grouping$invokeStatic (query_processor_test.clj:1494)
           metabase.query_processor_test$count_of_grouping$doInvoke (query_processor_test.clj:1493)
           on (query_processor_test.clj:1509)
           metabase.test.data.datasets$do_with_engine$invokeStatic (datasets.clj:72)
           on (datasets.clj:69)
           user$eval83552$invokeStatic (form-init8592673036465478053.clj:1)
           on (form-init8592673036465478053.clj:1)
```

I think you can't use the `k/raw` function on plain clojure jdbc statements. Maybe you know how to do that properly?

camsaul commented on src/metabase/driver/crate/native.clj:
```diff
@@ -0,0 +1,31 @@
+                                         :columns columns
+                                         :cols    (for [[column first-value] (partition 2 (interleave columns first-row))]
+                                                    {:name      column
+                                                     :base_type (n/value->base-type first-value)})})))))
```
This should be with indented like 

```clojure
(jdbc/with-db-connection [t-conn db-conn]
  (let [^java.sql.Connection jdbc-connection (:connection t-conn)]
    ...))
```

camsaul commented on src/metabase/driver/crate.clj:
```diff
@@ -0,0 +1,97 @@
+         (dissoc opts :hosts)))
+
+(defn- connection-details->spec [_ details]
+  (-> details crate-spec))
```
The `->` threading macro isn't doing much here. Just write it as `(crate-spec details)`

camsaul commented on project.clj:
```diff
@@ -66,7 +66,10 @@
-                 [swiss-arrows "1.0.0"]]                              ; 'Magic wand' macro -<>, etc.
+                 [swiss-arrows "1.0.0"]                               ; 'Magic wand' macro -<>, etc.
+                 [io.crate/crate-jdbc "1.11.0"]
+                 [io.crate/crate-client "0.54.7"]]
```
Mind moving these up a bit? We keep all of Metabase's dependencies sorted by name. A comment explaining its purpose like the other dependencies would be helpful as well.

Commit:
use timezone specified in metabase settings

Commit:
set ip address field to unknown

Commit:
several code improvements

agilliland commented on src/metabase/driver/crate/util.clj:
```diff
@@ -12,8 +13,16 @@
-  "date_trunc('interval', timestamp): truncates a timestamp to a given interval"
-  (k/sqlfn :DATE_TRUNC (kx/literal unit) expr))
+  "date_trunc('interval', timezone, timestamp): truncates a timestamp to a given interval"
+  (if (or (nil? (driver/report-timezone)) (= (count (driver/report-timezone)) 0))
```
Accessing `driver/report-timezone` directly is not supported.  We provide access to any global settings information you might need directly on the query when it's passed in for execution, and in the case of the generic-sql driver implementation it creates a binding against `*query*` that you can easily access to get anything out of the query when you need it.

So you should be getting the report-timezone via `(get-in *query* [:settings :report-timezone])`

This is also more convenient because the value is fully normalized, so it will be nil if no timezone value has been set, otherwise it will be the user chosen timezone.

mikethebeer replied:
@agilliland understand. However, if I use it like that it always returns `nil` although I defined a report timezone in the settings. Accessing the setting directly like I did before worked well.

```clojure
  (let [timezone    (get-in qp/*query* [:settings :report-timezone])]
    (if (= (nil? timezone) true)  ;; this always returns nil
      (k/sqlfn :DATE_FORMAT format expr)
      (k/sqlfn :DATE_FORMAT format timezone expr)))
```

salsakran commented:
@chaudum what is the status of this PR?

agilliland commented:
This lgtm minus the signing of the CLA.

agilliland commented:
Looks like there are merge conflicts though, so you should update this branch and merge in the latest code from `master`

camsaul commented on src/metabase/driver/crate/native.clj:
```diff
@@ -0,0 +1,31 @@
+            [metabase.models.database :refer [Database]]
+            [metabase.db :refer [sel]]
+            [metabase.driver.generic-sql :as sql]
+            [clojure.tools.logging :as log]
```
do you mind putting the `clojure` `:requires` above the Metabase ones? We keep them ordered like:

1.  Clojure namespaces
2.  3rd-party namespaces
3.  Metabase namespaces

camsaul commented on src/metabase/driver/crate/generic_sql.clj:
```diff
@@ -0,0 +1,37 @@
+(ns metabase.driver.crate.generic-sql
+  (:require [metabase.driver.generic-sql :as sql]
+            [korma.core :as k]
```
Do you mind putting the `korma` require above the Metabase ones? We keep them ordered like:

1.  Clojure namespaces
2.  3rd-party namespaces
3.  Metabase namespaces

camsaul commented on src/metabase/driver/crate/query_processor.clj:
```diff
@@ -0,0 +1,30 @@
+            [metabase.driver.generic-sql.query-processor :as qp]
+            [korma.sql.fns :as kfns]
+            [korma.sql.engine :as kengine]
+            [metabase.query-processor.interface :as i])
```
`:require` ordering

camsaul commented on src/metabase/driver/crate/util.clj:
```diff
@@ -0,0 +1,84 @@
+            [korma.sql.utils :as kutils]
+            [korma.core :as k]
+            [metabase.util :as u]
+            [metabase.driver.generic-sql.query-processor :as qp])
```
`:require` ordering

camsaul commented on test/metabase/test/data/crate.clj:
```diff
@@ -0,0 +1,75 @@
+            (metabase.test.data [generic-sql :as generic]
+                                [interface :as i])
+            [clojure.tools.logging :as log]
+            [korma.core :as k])
```
`:require` ordering

camsaul commented on test/metabase/test/data/crate.clj:
```diff
@@ -0,0 +1,75 @@
+  (merge generic/DefaultsMixin
+         {:execute-sql!              generic/sequentially-execute-sql!
+          :field-base-type->sql-type (fn [_ base-type]
+                                       (field-base-type->sql-type base-type))
```
You should just use the util function `drop-first-arg` here which is a bit cleaner. So instead you should just do

```clojure
:field-base-type->sql-type (u/drop-first-arg field-base-type->sql-type)
```

chaudum committed:
changed order of :require items

chaudum committed:
simplified field base type function declaration

chaudum commented:
hi @camsaul @agilliland @salsakran 
I've addressed the last comment, rebased with current master and signed the CLA. I guess @mikethebeer and @bogensberger will also need to sign it right?

agilliland commented:
yeah, lets have everyone who made contributes go ahead and sign the CLA.

chaudum commented:
@agilliland  all CLAs are signed. should I squash the commits?

tlrobinson commented:
Woo, congrats!

However, looks like CI is failing on master now (something in the Clojure linters)

agilliland commented:
yeah, there were a couple lint things that we missed.  i'll fix those up real quick.

definite congrats though.  we appreciate the Crate team taking the time to put together this contribution!

salsakran commented:
Hurray!

Thanks @mikethebeer, @chaudum @aslanbekirov + @bogensberger for all your help here!
